### PR TITLE
Improve automated messaging

### PR DESCRIPTION
**Pull request by filipsch**
- Bite-size messages that are pasted together
- More meaningful defaults that help (see test_messaging.py)
- If manually specifying incorrect_msg in has_equal_x: do not prepend previous
- Get rid of default-generated messages scattered all over the place
- Better description of arguments
- Better description of calls
- Get rid of has_key and check_key docs, as they will be phased out
- Improved handling of getting results, output and values from process
- Simplify old test_expression_x functions and group in one file

@sumedh10 
- Would be good if you could give your opinion on the messages in the `test_messaging.py` file.
- Do you like the approach of not prepending the previous messages?

Currently if you want to tell people `"abc"` when they specify an argument wrong, you need:

```python
Ex().check_function("round").check_args(0).has_equal_value("abc")
# Produces: check the function round(). Check the first argument. abc.
Ex().check_function("round", expand_msg = "").check_args(0).has_equal_value("abc")
# Produces: abc.
```

With this change, since you manually specify the `incorrect_msg` in `has_equal_value()`, you no longer have to set `expand_msg` to empty string manually.

```python
Ex().check_function("round").check_args(0).has_equal_value("abc")
# Produces: abc.
```

This is true for any SCT chain that would use `has_equal_value/output/error()`. WDYT?

### CODE REVIEW

File: tests/test_messaging.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    if s and e:
+        return output['column_start'] == s and output['column_end'] == e
+    else:
+        return True
+
+@pytest.mark.parametrize('stu, patt, cols, cole', [
+        ('', 'Did you call `round()`?', None, None),
+        ('round(1)', 'Check your call of `round()`. Did you specify the second argument?', 1, 8),
+        ('round(1, a)', 'Check your call of `round()`. Did you correctly specify the second argument? Rerunning it generated an error: `name \'a\' is not defined`.', 10, 10),
+        ('round(1, 2)', 'Check your call of `round()`. Did you correctly specify the second argument? Expected `3`, but got `2`.', 10, 10)
```
Comment (sumedh10) [inline tests/test_messaging.py]: I am not in favor of the last statement. 

> Expected `3`, but got `2`. 

Not only here, but whenever we use `Ex().check_object().has_equal_value()` the default message gives away the answer. Can we end the message at "Did you correctly specify the second argument?"

(this is how we do it in `testwhat`)

File: tests/test_messaging.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+def lines(output, s, e):
+    if s and e:
+        return output['column_start'] == s and output['column_end'] == e
+    else:
+        return True
+
+@pytest.mark.parametrize('stu, patt, cols, cole', [
+        ('', 'Did you call `round()`?', None, None),
+        ('round(1)', 'Check your call of `round()`. Did you specify the second argument?', 1, 8),
+        ('round(1, a)', 'Check your call of `round()`. Did you correctly specify the second argument? Rerunning it generated an error: `name \'a\' is not defined`.', 10, 10),
```
Comment (sumedh10) [inline tests/test_messaging.py]: Not sure why we say "Rerunning" here.

File: tests/test_messaging.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        'DC_SCT': 'Ex().check_object("x").has_equal_value()'
+    })
+    assert not output['correct']
+    assert message(output, patt)
+    assert lines(output, cols, cole)
+
+@pytest.mark.parametrize('stu, patt', [
+    ('', 'The system wants to check the definition of `test()` but hasn\'t found it.'),
+    ('def test(a, b): return 1', 'Check the definition of `test()`. Calling `test(1, 2)` should return `3`, instead got `1`.'),
+    ('def test(a, b): return a + b', 'Check the definition of `test()`. Calling `test(1, 2)` should output `3`, instead got no output.'),
```
Comment (sumedh10) [inline tests/test_messaging.py]: Line 61 says "should return" and line 62 says "should output".

File: tests/test_messaging.py (hunk 2 above)
Comment (filipsch) [reply tests/test_messaging.py]: Good point. updated.

File: tests/test_messaging.py (hunk 3 above)
Comment (filipsch) [reply tests/test_messaging.py]: Line 61 is failing on the `call(..., 'value')` function, so talks about what the function should return.
Line 62 is failing on the `call(..., 'output')` function, so talks about what the function should print out.
The messaging for output should be improved to be refer to print out. Thanks for flagging.

File: tests/test_messaging.py (hunk 1 above)
Comment (filipsch) [reply tests/test_messaging.py]: Unfortunately, the architecture of `pythonwhat` is slightly more complicated than this. If OK with you, I will not tackle this in this PR yet. When I can lock out some time for this, I'd be happy to discuss this with you in more detail before I make any adjustments. OK?

Comment (sumedh10) [reply tests/test_messaging.py]: Absolutely! As long as you agree :D

**Comment by sumedh10** [comment]
@filipsch Although I mentioned this on slack, I want to say this again. Huge fan of not having to manually specify `expand_msg = " "`.

**Pull request merged**